symengine/bottom_up.cpp

```diff
@@ -0,0 +1,139 @@
+#include <symengine/visitor.h>
+
+namespace SymEngine
+{
+
+inline RCP<const Basic> identity(const RCP<const Basic> &arg)
+{
+    return arg;
+}
+
+template <typename Func>
+class BottomUpVisitor : public BaseVisitor<BottomUpVisitor<Func>>
+{
+protected:
+    RCP<const Basic> result_;
+    Func f;
+
+public:
+    BottomUpVisitor(const Func &f_) : f(f_)
+    {
+    }
+
+    RCP<const Basic> apply(const Basic &x)
+    {
+        return apply(x.rcp_from_this());
+    }
+
+    RCP<const Basic> apply(const RCP<const Basic> &x)
+    {
+        x->accept(*this);
+        return result_;
+    }
+
+    void bvisit(const Basic &x)
+    {
+        result_ = x.rcp_from_this();
+    }
+
+    void bvisit(const Add &x)
+    {
+        vec_basic newargs;
+        for (const auto &a : x.get_args()) {
+            newargs.push_back(apply(a));
+        }
+        result_ = f(add(newargs));
```

Comment:
Nice !!
I will remove this from here and other bvisits as well except `bvisit(Basic &x)`.